// node-html-to-image/dist/models/Screenshot.js
"use strict";
exports.__esModule = true;
exports.Screenshot = void 0;
var Screenshot = /** @class */ (function () {
    function Screenshot(params) {
        if (!params || !params.html) {
            throw Error("You must provide an html property.");
        }
        var html = params.html, encoding = params.encoding, _a = params.transparent, transparent = _a === void 0 ? false : _a, output = params.output, content = params.content, _b = params.selector, selector = _b === void 0 ? "body" : _b, _c = params.quality, quality = _c === void 0 ? 80 : _c, _d = params.type, type = _d === void 0 ? "png" : _d;
        this.html = html;
        this.encoding = encoding;
        this.transparent = transparent;
        this.type = type;
        this.output = output;
        this.content = isEmpty(content) ? undefined : content;
        this.selector = selector;
        this.quality = type === "jpeg" ? quality : undefined;
    }
    Screenshot.prototype.setHTML = function (html) {
        if (!html) {
            throw Error("You must provide an html property.");
        }
        this.html = html;
    };
    Screenshot.prototype.setBuffer = function (buffer) {
        this.buffer = buffer;
    };
    return Screenshot;
}());
exports.Screenshot = Screenshot;
function isEmpty(val) {
    return val == null || !Object.keys(val).length;
}
